components/core/src/clp/aws/AwsAuthenticationSigner.cpp

```diff
@@ -0,0 +1,337 @@
+                __FILENAME__,
+                __LINE__,
+                "Invalid S3 HTTP URL format: " + url
+        );
+    }
+
+    if (m_region.empty()) {
+        m_region = cDefaultRegion;
+    }
+    m_host = fmt::format("{}.s3.{}.amazonaws.com", m_bucket, m_region);
+}
+
+auto AwsAuthenticationSigner::generate_presigned_url(
+        S3Url const& s3_url,
+        string& presigned_url
+) const -> ErrorCode {
+    auto const s3_region = s3_url.get_region();
+
+    auto const now = std::chrono::system_clock::now();
+    auto const date_string = get_date_string(now);
+    auto const timestamp_string = get_timestamp_string(now);
+
+    auto const scope = get_scope(date_string, s3_region);
+    auto const canonical_query_string = generate_canonical_query_string(scope, timestamp_string);
+
+    auto const canonical_request = get_canonical_request(s3_url, canonical_query_string);
+
+    string string_to_sign{};
+    if (auto error_code
+        = get_string_to_sign(scope, timestamp_string, canonical_request, string_to_sign);
+        ErrorCode_Success != error_code)
+    {
+        return error_code;
+    }
+
+    vector<unsigned char> signature{};
+    if (auto error_code = get_signature(s3_region, date_string, string_to_sign, signature);
+        ErrorCode_Success != error_code)
+    {
+        return error_code;
+    }
+    auto const signature_str = convert_to_hex_string({signature.data(), signature.size()});
+
+    presigned_url = fmt::format(
+            "https://{}{}?{}&{}={}",
+            s3_url.get_host(),
+            s3_url.get_path(),
+            canonical_query_string,
+            cXAmzSignature,
+            signature_str
+    );
+    return ErrorCode_Success;
+}
+
+auto AwsAuthenticationSigner::get_signature(
+        string_view region,
+        string_view date_string,
+        string_view string_to_sign,
+        vector<unsigned char>& signature
+) const -> ErrorCode {
+    vector<unsigned char> signing_key{};
+    if (auto error_code = get_signing_key(region, date_string, signing_key);
+        ErrorCode_Success != error_code)
+    {
+        return error_code;
+    }
+
+    if (auto error_code = get_hmac_sha256_hash(
+                {size_checked_pointer_cast<unsigned char const>(string_to_sign.data()),
+                 string_to_sign.size()},
+                {size_checked_pointer_cast<unsigned char const>(signing_key.data()),
+                 signing_key.size()},
+                signature
+        );
+        ErrorCode_Success != error_code)
+    {
+        return error_code;
+    }
+    return ErrorCode_Success;
+}
+
+auto AwsAuthenticationSigner::get_signing_key(
+        string_view region,
+        string_view date_string,
+        vector<unsigned char>& signing_key
+) const -> ErrorCode {
+    string key{cAws4};
+    key += m_secret_access_key;
+
+    vector<unsigned char> date_key{};
+    vector<unsigned char> date_region_key{};
+    vector<unsigned char> date_region_service_key{};
```

Comment:
```suggestion
    vector<unsigned char> date_key;
    vector<unsigned char> date_region_key;
    vector<unsigned char> date_region_service_key;
```
Also, how about moving the declaration before the if statement that they are firstly used?